src/frontend/qt_sdl/LuaMain.cpp

```diff
@@ -0,0 +1,583 @@
+    return 0;
+}
+AddLuaFunction(lua_MelonClear,MelonClear);
+
+enum ramInfo_ByteType
+{
+    ramInfo_OneByte = 1, 
+    ramInfo_TwoBytes = 2, 
+    ramInfo_FourBytes = 4
+};
+
+u32 GetMainRAMValueU(const u32& addr, const ramInfo_ByteType& byteType)
+{
+    switch (byteType)
+    {
+    case ramInfo_OneByte:
+        return *(u8*)(NDS::MainRAM + (addr&NDS::MainRAMMask));
+    case ramInfo_TwoBytes:
+        return *(u16*)(NDS::MainRAM + (addr&NDS::MainRAMMask));
+    case ramInfo_FourBytes:
+        return *(u32*)(NDS::MainRAM + (addr&NDS::MainRAMMask));
+    default:
+        return 0;
+    }
+}
+
+s32 GetMainRAMValueS(const u32& addr, const ramInfo_ByteType& byteType)
+{
+    switch (byteType)
+    {
+    case ramInfo_OneByte:
+        return *(s8*)(NDS::MainRAM + (addr&NDS::MainRAMMask));
+    case ramInfo_TwoBytes:
+        return *(s16*)(NDS::MainRAM + (addr&NDS::MainRAMMask));
+    case ramInfo_FourBytes:
+        return *(s32*)(NDS::MainRAM + (addr&NDS::MainRAMMask));
+    default:
+        return 0;
+    }
+}
+
+
+int Lua_ReadDatau(lua_State* L,ramInfo_ByteType byteType) 
+{   
+    u32 address = luaL_checkinteger(L,1);
+    u32 value =GetMainRAMValueU(address,byteType);
+    lua_pushinteger(L, value);
+    return 1;
+}
+
+int Lua_ReadDatas(lua_State* L,ramInfo_ByteType byteType)
+{
+    u32 address = luaL_checkinteger(L,1);
+    s32 value = GetMainRAMValueS(address,byteType);
+    lua_pushinteger(L, value);
+    return 1;
+}
+
+int Lua_Readu8(lua_State* L)
+{
+    return Lua_ReadDatau(L,ramInfo_OneByte);
+}
+AddLuaFunction(Lua_Readu8,Readu8);
+
+int Lua_Readu16(lua_State* L)
+{
+    return Lua_ReadDatau(L,ramInfo_TwoBytes);
+}
+AddLuaFunction(Lua_Readu16,Readu16);
+
+int Lua_Readu32(lua_State* L)
+{
+    return Lua_ReadDatau(L,ramInfo_FourBytes);
+}
+AddLuaFunction(Lua_Readu32,Readu32);
+
+int Lua_Reads8(lua_State* L)
+{
+    return Lua_ReadDatas(L,ramInfo_OneByte);
+}
+AddLuaFunction(Lua_Reads8,Reads8);
+
+int Lua_Reads16(lua_State* L)
+{
+    return Lua_ReadDatas(L,ramInfo_TwoBytes);
+}
+AddLuaFunction(Lua_Reads16,Reads16);
+
+int Lua_Reads32(lua_State* L)
+{
+    return Lua_ReadDatas(L,ramInfo_FourBytes);
+}
+AddLuaFunction(Lua_Reads32,Reads32);
+
+
+int Lua_NDSTapDown(lua_State* L)
+{
+    int x =luaL_checkinteger(L,1);
+    int y =luaL_checkinteger(L,2);
+    NDS::TouchScreen(x,y);
+    
+    return 0;
+}
+AddLuaFunction(Lua_NDSTapDown,NDSTapDown);
+
+int Lua_NDSTapUp(lua_State* L)
+{
+    NDS::ReleaseScreen();
+    return 0;
+}
+AddLuaFunction(Lua_NDSTapUp,NDSTapUp);
+
+int Lua_stateSave(lua_State* L)
+{
+    QString filename = luaL_checkstring(L,1);
+    luaThread->luaStateSave(filename);
+    return 0;
+}
+AddLuaFunction(Lua_stateSave,StateSave);
+
+int Lua_stateLoad(lua_State* L)
+{
+    QString filename = luaL_checkstring(L,1);
+    luaThread->luaStateLoad(filename);
+    return 0;
+}
+AddLuaFunction(Lua_stateLoad,StateLoad);
+
+
+int Lua_nextFrame(lua_State* L)
+{
+    frameFlag=false;
+    while(!frameFlag)
```

Comment:
frameFlag is not volatile, so the compiler will probably skip the loop entirely (and if it doesn't this'll burn a lot of cycles).